lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -816,3 +829,243 @@ def get_provider_vault_token(
+
+        Returns:
+            The CertificateRequestAttributes that should be used to request the
+            intermediate CA certificate from the TLS certificates provider.
+        """
+        return CertificateRequestAttributes(
+            self._common_name,
+            is_ca=True,
+        )
+
+    def validate_input(self, *, common_name: str):
+        """Validate the inputs to this class."""
+        if common_name == "":
+            raise ManagerError("Common name cannot be empty.")
+
+    def _get_pki_intermediate_ca_from_relation(
+        self,
+    ) -> tuple[ProviderCertificate | None, PrivateKey | None]:
+        """Get the intermediate CA certificate and private key from the relation data.
+
+        This is the CA certificate that the provider charm has issued to Vault.
+        """
+        provider_certificate, private_key = self._tls_certificates_pki.get_assigned_certificate(
+            certificate_request=self.intermediate_ca_cert_request()
+        )
+        if not provider_certificate:
+            logger.debug("No intermediate CA certificate available")
+        if not private_key:
+            logger.debug("No private key available")
+        return provider_certificate, private_key
+
+    def _get_vault_service_ca_certificate(self) -> Certificate | None:
+        """Get the current CA certificate from the Vault service."""
+        try:
+            intermediate_ca_cert = self._vault_client.get_intermediate_ca(mount=self._mount_point)
+            if not intermediate_ca_cert:
+                return None
+            return Certificate.from_string(intermediate_ca_cert)
+        except (VaultClientError, TLSCertificatesError) as e:
+            logger.error("Failed to get current CA certificate: %s", e)
+            return None
+
+    def _intermediate_ca_exceeds_role_ttl(self, intermediate_ca_certificate: Certificate) -> bool:
+        """Check if the intermediate CA's remaining validity exceeds the role's max TTL.
+
+        Vault PKI enforces that issued certificates cannot outlast their signing CA.
+        This method ensures that the intermediate CA's remaining validity period
+        is longer than the maximum TTL allowed for certificates issued by this role.
+        """
+        current_ttl = self._vault_client.get_role_max_ttl(
+            role=self._role_name, mount=self._mount_point
+        )
+        if (
+            not current_ttl
+            or not intermediate_ca_certificate.expiry_time
+            or not intermediate_ca_certificate.validity_start_time
+        ):
+            return False
+        certificate_validity = (
+            intermediate_ca_certificate.expiry_time
+            - intermediate_ca_certificate.validity_start_time
+        )
+        certificate_validity_seconds = certificate_validity.total_seconds()
+        return certificate_validity_seconds > current_ttl
+
+    def configure(self):
+        """Perform the initial configuration of the PKI backend."""
```

Comment:
Not just initial right? And I wouldn't make assumptions as to when this method is called (as it's outside of this function's control), I would simply encourage to state what it does.

Reply:
Yeah this was a bad one. I put it in as a placeholder, because there was nothing in the original function it was moved from, and then never revisited it.

I'll write something a little better here.

Reply:
https://github.com/canonical/vault-k8s-operator/pull/559/commits/75059d5fc72815ba4e8ea54a0eba1c40bdb5704f